### PR TITLE
Simplify DictionaryBuilder constructors (#2684) (#2054)

### LINKED ISSUE #2684: DictionaryBuilders can Create Invalid DictionaryArrays
**Describe the bug**

Both `StringDictionaryBuilder` and `PrimitiveDictionaryBuilder` are created with existing builders. This is problematic if the key builders are not empty, as these already inserted keys will not be validated

**To Reproduce**

**Expected behavior**

Requiring the builders to create a dictionary builder is a pretty strange API, and I would like to propose just removing it

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2684
Part of #2054

# Rationale for this change
 

The previous constructors were confusing, unnecessarily verbose, and allowed creating invalid DictionaryArrays by passing in a non-empty keys builder

# What changes are included in this PR?

# Are there any user-facing changes?

Yes

### CODE REVIEW

File: arrow/src/array/builder/string_dictionary_builder.rs [hunk 1]
```diff
@@ -98,6 +107,24 @@ where
...
+    ) -> Self {
+        Self {
+            state: Default::default(),
+            dedup: Default::default(),
+            keys_builder: PrimitiveBuilder::with_capacity(keys_capacity),
+            values_builder: StringBuilder::with_capacity(value_capacity, string_capacity),
+        }
+    }
+
     /// Creates a new `StringDictionaryBuilder` from a keys builder and a dictionary
```
Comment (viirya) [inline arrow/src/array/builder/string_dictionary_builder.rs:128]:
```suggestion
    /// Creates a new `StringDictionaryBuilder` from a keys capacity and a dictionary
```

File: arrow/src/array/builder/primitive_dictionary_builder.rs [hunk 2]
```diff
@@ -95,22 +93,41 @@ where
...
-    pub fn new(
-        keys_builder: PrimitiveBuilder<K>,
-        values_builder: PrimitiveBuilder<V>,
-    ) -> Self {
+    pub fn new() -> Self {
         Self {
-            keys_builder,
-            values_builder,
+            keys_builder: PrimitiveBuilder::new(),
+            values_builder: PrimitiveBuilder::new(),
```
Comment (viirya) [inline arrow/src/array/builder/primitive_dictionary_builder.rs:115]: I just wonder if the issue is provided key builder might be not empty, we can check the length of the key builder when it is passed in? The API could be unchanged.

File: arrow/src/array/builder/primitive_dictionary_builder.rs [hunk 3]
```diff
@@ -95,22 +93,41 @@ where
...
+        Self::new()
+    }
+}
+
 impl<K, V> PrimitiveDictionaryBuilder<K, V>
 where
     K: ArrowPrimitiveType,
     V: ArrowPrimitiveType,
 {
     /// Creates a new `PrimitiveDictionaryBuilder` from a keys builder and a value builder.
```
Comment (viirya) [inline arrow/src/array/builder/primitive_dictionary_builder.rs:111]:
```suggestion
    /// Creates a new `PrimitiveDictionaryBuilder`.
```

File: arrow/src/array/builder/primitive_dictionary_builder.rs (hunk 2 above)
Comment (tustvold) [reply arrow/src/array/builder/primitive_dictionary_builder.rs:115]: For sure, however, I'd argue it is a good opportunity to bring this API closer into alignment with the other builders. We don't require passing the various different inner builders to StringArray, for example. The only other place we do is ListArray, which again I think is a little odd.

File: arrow/src/array/builder/string_dictionary_builder.rs (hunk 1 above)
Comment (viirya) [inline arrow/src/array/builder/string_dictionary_builder.rs:128]:
```suggestion
    /// Creates a new `StringDictionaryBuilder` from a keys capacity and a dictionary
```

**Review by viirya — APPROVED** [review]

File: arrow/src/array/builder/primitive_dictionary_builder.rs (hunk 2 above)
Comment (viirya) [reply arrow/src/array/builder/primitive_dictionary_builder.rs:115]: I see. Yea, it is somehow weird to provide key and value builders from outside of these dictionary builder.

**Pull request merged**